vent/coordinator/control_module.py

```diff
@@ -221,9 +198,50 @@ def update(self, pressure):
             if self.pressure < self.PEEP:
                 self.Qin = 1
         else:
-            self.cycle_start = time.time() # restart
+            self.cycle_start    = time.time() # new cycle starts
+            self.cycle_counter += 1
+
+        if self.cycle_counter not in self.cycle_pressure_waveforms.keys():   #if this cycle doesn't exist yet, start it
+            self.cycle_pressure_waveforms[self.cycle_counter] = np.array([[0, pressure]])
+        else:
+            data = self.cycle_pressure_waveforms[self.cycle_counter]
+            data = np.append(data, [[cycle_phase, pressure]], axis=0)
+            self.cycle_pressure_waveforms[self.cycle_counter] = data
 
 
+def test_critical_levels(min, max, value, name, active_alarms, logged_alarms):
```

Comment:
One other suggestion, maybe test_critical_levels could be a method of the ControlModuleBase class? And active_alarms and logged_alarms could be variables of that class? Then you wouldn't need to pass the alarm data structures around.

Reply:
Yes! That would make sense, but it adds to the architecture. I didn't want to touch that.

Reply:
I think the architecture should evolve as a mix of top-down and bottom-up contributions. So please add things as they make sense and we can update the arch spec to reflect that.

Reply:
alright, will do that tonight

Reply:
...like this? 

[Also also Hanna H says hi and misses the times with you as office mate :-)]

Reply:
If ok, can you approve the pull request? Then jonny can merge the new UI code with these updates.